### DISCUSSION TITLE
File size (or content size) limitations?

### DISCUSSION THREAD

**Discussion by aorsten** [exceljs/exceljs#2568] [Q&A]
Hi, thanks for the great work on exceljs!
I'm using it in the browser, and am loading files using the buffer load option. 

```
// load from buffer
const workbook = new Excel.Workbook();
await workbook.xlsx.load(data);
// ... use workbook
```

I just came across an Excel file of 36MB size with about 25 worksheets, where at least 5 of the sheets have between 10-20k rows of data in about 5-8 columns. It never stops loading. I removed all but 1 sheet with 15k rows of data, and still loading it takes about 12s.

So, that leads to some questions about ExcelJS in the browser:
 1. Are there any known limitations to practical file sizes or amount of lines/sheets? 
 2. Is there any way to read the progress of loading, to learn whether it's getting anywhere and how far loading has come?
 3. Is there any way to parse the line count before loading, to learn whether the file is too large?
 4. Has anyone tried using web workers to offload the memory heavy process of a big file? Does it improve things?

**Comment by Sarsewar** [comment]
@aorsten Did you find anything related to your questions? Please share your analysis.

**Reply by aorsten** [reply]
I'm sorry, but I didn't get anywhere with it yet.

**Comment by aorsten** [comment]
I am digging further into this issue now. There is a function called Worksheet._parseMergeCells() which is taking a long time. I inspected the workbook, and yes - the sheets have merged cells in every single row. 

So the way I see it, ExcelJS (at least in the browser) is too slow for this scenario. I think ideally I would like there to be some way to pre-detect this, so I can warn my users about it.

**Reply by aorsten** [reply]
I am noticing now that these lines are the slow ones:

```
// worksheet.js, ._mergeCellsInternal
    // check cells aren't already merged
    _.each(this._merges, merge => {
      if (merge.intersects(dimensions)) {
        throw new Error('Cannot merge already merged cells');
      }
    });
```
If I comment out these lines, parsing is much faster. It seems to me that ExcelJs revisits every single previous merge to prevent any of them from overlapping, which - when there are 10000 merges in a worksheet - increases the loops to n^2. 

I am curious whether this could be made more efficient?

**Reply by aorsten** [reply]
I created an issue for this: https://github.com/exceljs/exceljs/issues/2689
